rust/noosphere/src/ffi/fs.rs

```diff
@@ -0,0 +1,233 @@
+use anyhow::anyhow;
+use noosphere_core::data::Did;
+use noosphere_fs::{SphereFile, SphereFs};
+use safer_ffi::prelude::*;
+use std::{pin::Pin, str::FromStr};
+use subtext::{Peer, Slashlink};
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+use crate::{
+    ffi::{NsHeaders, NsNoosphereContext},
+    platform::{PlatformKeyMaterial, PlatformStore},
+};
+
+ReprC! {
+    #[ReprC::opaque]
+    pub struct NsSphereFs {
+        inner: SphereFs<PlatformStore, PlatformKeyMaterial>
+    }
+}
+
+impl NsSphereFs {
+    pub fn inner(&self) -> &SphereFs<PlatformStore, PlatformKeyMaterial> {
+        &self.inner
+    }
+
+    pub fn inner_mut(&mut self) -> &mut SphereFs<PlatformStore, PlatformKeyMaterial> {
+        &mut self.inner
+    }
+}
+
+ReprC! {
+    #[ReprC::opaque]
+    pub struct NsSphereFile {
+        inner: SphereFile<Pin<Box<dyn AsyncRead>>>
+    }
+}
+
+impl NsSphereFile {
+    pub fn inner(&self) -> &SphereFile<Pin<Box<dyn AsyncRead>>> {
+        &self.inner
+    }
+
+    pub fn inner_mut(&mut self) -> &mut SphereFile<Pin<Box<dyn AsyncRead>>> {
+        &mut self.inner
+    }
+}
+
+#[ffi_export]
+/// Initialize an instance of a [NsSphereFs] that is a read/write view into
+/// the contents (as addressed by the slug namespace) of the identitifed sphere
+pub fn ns_sphere_fs_open(
+    noosphere: &NsNoosphereContext,
+    sphere_identity: char_p::Ref<'_>,
+) -> repr_c::Box<NsSphereFs> {
+    noosphere
+        .async_runtime()
+        .block_on(async {
+            let sphere_context = noosphere
+                .inner()
+                .get_sphere_context(&Did(sphere_identity.to_str().into()))
+                .await?;
+
+            let sphere_context = sphere_context.lock().await;
+
+            Ok(repr_c::Box::new(NsSphereFs {
+                inner: sphere_context.fs().await?,
+            })) as Result<_, anyhow::Error>
+        })
+        .unwrap()
+}
+
+#[ffi_export]
+/// De-allocate an [NsSphereFs] instance
+pub fn ns_sphere_fs_free(sphere_fs: repr_c::Box<NsSphereFs>) {
+    drop(sphere_fs)
+}
+
+#[ffi_export]
+/// Read a [NsSphereFile] from a [NsSphereFs] instance by slashlink. Note that
+/// although this function will eventually support slashlinks that include the
+/// pet name of a peer, at this time only slashlinks with slugs referencing the
+/// slug namespace of the local sphere are allowed.
+pub fn ns_sphere_fs_read(
+    noosphere: &NsNoosphereContext,
+    sphere_fs: &NsSphereFs,
+    slashlink: char_p::Ref<'_>,
+) -> Option<repr_c::Box<NsSphereFile>> {
+    noosphere
+        .async_runtime()
+        .block_on(async {
+            let slashlink = Slashlink::from_str(slashlink.to_str())?;
+
+            if Peer::None != slashlink.peer {
```

Comment:
That's an interesting idea. Worth noting:

We are guaranteed to take a dependency on `subtext` until such time as we are able to define "generic" content codecs; in other words, this is a pragmatic necessity for Noosphere for the time being.

Even when we are truly generic over content type, we are very likely to move all links around as some Slashlink-like construct (which is essentially our form of URL, pairing peer name and slug and possibly other things). In other words, part of what a content codec will do is tell us about all the slashlinks in a piece of content.

Consider:

 - What if `fetch` asked you to break apart a URL into its constituent parts before making a request?
 - A peer name only has absolute meaning in the context of a sphere

Reply:
> We are guaranteed to take a dependency on subtext until such time as we are able to define "generic" content codecs; in other words, this is a pragmatic necessity for Noosphere for the time being.

Sounds good! I'd be curious to learn more about these codecs, if implemented anywhere currently. Lots of interesting extension opportunities here, personally for something like Markdown, but also as a basis for e.g. a self-referencing wikipedia clone.

> Even when we are truly generic over content type, we are very likely to move all links around as some Slashlink-like construct (which is essentially our form of URL, pairing peer name and slug and possibly other things). In other words, part of what a content codec will do is tell us about all the slashlinks in a piece of content.

I understand some `SlashLink` struct needed throughout internally; no opinion on if this is exposed via FFI (`NsLink`), or primitives are handled directly by methods (e.g. two `char*`s).

> What if fetch asked you to break apart a URL into its constituent parts before making a request?

Is `ns_sphere_fs_read` == `fetch` in this context? Handling the parsing of e.g. `@cdata/dogs` in the embedder (as either NsLink or two chars) would align with a codec-agnostic future. Again no pref between `NsLink`/two `char*`s

Reply:
> I'd be curious to learn more about these codecs, if implemented anywhere currently

This is very speculative and not currently implemented. You can find the stand-in issue here: https://github.com/subconsciousnetwork/noosphere/issues/50

Let's discuss in detail there if you are interested!

> I understand some SlashLink struct needed throughout internally; no opinion on if this is exposed via FFI (NsLink), or primitives are handled directly by methods (e.g. two char*s).

Yah, it may be that we need a more generic notion of slashlink that we include here in the core crates.

> Is ns_sphere_fs_read == fetch in this context?

Yes. I'm trying to think of FFI in terms of what data is the embedder likely to have on-hand. If it is merely performance for them to convert a string to an opaque struct pointer when they would be doing that every time, why not just pass a string?

`fetch` takes a URL string and handles the rest for you.